### PR TITLE
Eslint 9 migration

### LINKED ISSUE #230339: Adopt eslint flat configs
Tracks adopting eslint's flat configuration setup so that we can update to eslint 9+

### PR DESCRIPTION
**Pull request by mjbvz**
Fixes #230339

Migrates us to eslint 9 and a flat config 

TODO:

- [X] Migrate main eslint 
- [X] Fix local rules
- [x] Convert remaining `eslintrc` files 
- [x] Fix all lint errors
- [x] Validate build scripts still work
- [x] Validate editor tooling still works

Follow ups:

- [ ] All local rules should have proper schemas (worked around by setting `schema: false`
- [ ] Fix the remaining unsafe casts (or really leave them suppressed)

### CODE REVIEW

File: src/vs/amdX.ts [hunk 1]
```diff
@@ -70,10 +70,8 @@ class AMDModuleImporter {
 		(globalThis as any).define.amd = true;
 
 		if (this._isRenderer) {
-			// eslint-disable-next-line no-restricted-globals
 			this._amdPolicy = (globalThis as any)._VSCODE_WEB_PACKAGE_TTP ?? window.trustedTypes?.createPolicy('amdLoader', {
```
Comment (mjbvz) [inline src/vs/amdX.ts:73]: @bpasero I removed these as I don't see any rules that would apply to them anymore. In the new eslint version, having a suppression comment that does nothing is an error

 Let me know if this is incorrect and we still should be getting lint errors here

File: src/vscode-dts/vscode.d.ts [hunk 2]
```diff
@@ -7505,7 +7506,7 @@ declare module 'vscode' {
 		 * must be activated. Check whether {@link TerminalShellExecution.exitCode} is rejected to
 		 * verify whether it was successful.
 		 *
-		 * @param command A command to run.
+		 * @param executable A command to run.
```
Comment (mjbvz) [inline src/vscode-dts/vscode.d.ts:7509]: Not sure why these weren't caught before. All seem like real lint errors that should have been reported

File: src/vs/amdX.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/amdX.ts:73]: @mjbvz but the rule is in use and actually catches very ugly uses of globals such as `event` or `length`:

https://github.com/microsoft/vscode/blob/de557e7b66cc8a085ae80f6b3f1e4b496a1f6e6c/.eslintrc.json#L39-L50

Comment (bpasero) [reply src/vs/amdX.ts:73]: Or are you saying, the rule is still there but just does not apply to this one file `amdX.ts`? That can well be true.

**Comment by sandy081** [comment]
@mjbvz Pushed a change to use `allowImportNames` for one of the rules I have created.

File: src/vs/amdX.ts (hunk 1 above)
Comment (mjbvz) [reply src/vs/amdX.ts:73]: Yes sorry this was unclear. The rule still seems to work but no lint error is reported on this specific line. I just wanted to double check with you that that was ok. Maybe `window` or `globalThis` was in the list of restricted names at some point?

**Comment by mjbvz** [comment]
Checks look good. Build script pass and I tested introducing a few errors and confirming they were reported. Planning to merge but please keep an eye out for any regressions or cases where our custom rules no longer apply

**Review by lramos15 — DISMISSED** [review]

**Comment by mjbvz** [comment]
cc @jrieken. Need your review as code owner as I touched `vscode.d.ts` to fix some new linting errors that were discovered

File: eslint.config.mjs [hunk 3]
```diff
@@ -0,0 +1,1393 @@
...
+					]
+				}
+			]
+		}
+	},
+	// Terminal
+	{
+		files: [
+			'src/vs/workbench/contrib/terminal/**/*.ts'
+		],
```
Comment (Tyriar) [inline eslint.config.mjs:1319]: Can you add `'src/vs/workbench/contrib/terminalContrib/**/*.ts'` here too? You'll also need to delete it at https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/terminalContrib/.eslintrc.json

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/amdX.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/amdX.ts:73]: I think its more likely that `amdX` is not covered by ESlint rules anymore or for a while now given its rather unusual location in `src/vs`.

**Review by jrieken — COMMENTED** [review]
:+1: for changes in `src/vscode-dts/vscode.d.ts`

File: eslint.config.mjs (hunk 3 above)
Comment (mjbvz) [reply eslint.config.mjs:1319]: Ah sorry, yes missed that we had some `eslintrc.json` too. Will make sure those also get ported

Comment (jrieken) [reply eslint.config.mjs:1319]: @Tyriar why do we have component specific eslint rules? Isn't the whole idea that we have team agreement and can feel at home anywhere?

**Comment by mjbvz** [comment]
Fixed merge conflicts and pushed a few small fixes

@jrieken If the changes look good, can you please approve the pr through a GitHub review too. It seems like our code owner rules require that either you or I approve any changes to `vscode.d.ts`

File: eslint.config.mjs (hunk 3 above)
Comment (Tyriar) [reply eslint.config.mjs]: DMed you

**Review by Tyriar — APPROVED** [review]
My parts all good

**Review by jrieken — APPROVED** [review]

**Pull request merged**